### PR TITLE
Use prebuilt binaries for cairo compilation

### LINKED ISSUE #17: Replace cairo compiler installation from source with prebuilt binaries
Instead of using rustc and building the cairo compiler from the git repo (from source), we could rely on just fetching and extracting the Assets in each release: https://github.com/starkware-libs/cairo/releases. This would significantly improve our build time.

Keep in mind that we build images for two architectures: arm64 and amd64 (x86). Probably one image will need one asset.

### PR DESCRIPTION
**Pull request by Nathan-SL**
Closes #17

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -21,9 +21,10 @@ jobs:
           command: ./build_and_push.sh
           environment:
             TAG_SUFFIX: ""
-            CAIRO_VERSION: 0.11.2
+            CAIRO_VERSION: "v0.11.2"
             OZ_VERSION: 0.5.1
             CAIRO_COMPILER_TARGET_TAG: "v1.1.0"
+            ASSET_NAME: "release-x86_64-unknown-linux-musl.tar.gz"
```
Comment (FabijanC) [inline .circleci/config.yml]: Rename `ASSET_NAME` to `CAIRO_COMPILER_ASSET_NAME`

File: Dockerfile [hunk 2]
```diff
@@ -10,11 +9,12 @@ RUN apk add git musl-dev curl
...
 
-# Install cairo1 compiler
-RUN git clone -b $CAIRO_COMPILER_TARGET_TAG https://github.com/starkware-libs/cairo.git
+ARG COMPILER_BINARY_URL
+ARG ASSET_NAME
 
-RUN cargo build --release --manifest-path /cairo/crates/cairo-lang-starknet/Cargo.toml
-RUN cargo build --release --manifest-path /cairo/crates/cairo-lang-sierra-to-casm/Cargo.toml
+# Download cairo1 compiler
+RUN curl --location -O --request GET $COMPILER_BINARY_URL
```
Comment (FabijanC) [inline Dockerfile]: Instead of relying on `curl`, we can rely on Dockerfile's `ADD` (which works with URLs)

File: Dockerfile [hunk 3]
```diff
@@ -10,11 +9,12 @@ RUN apk add git musl-dev curl
 ARG SHELL=/bin/ash
 RUN curl --proto '=https' --tlsv1.2 -sSf https://docs.swmansion.com/scarb/install.sh | $SHELL -s -- -v $SCARB_VERSION
 
-# Install cairo1 compiler
-RUN git clone -b $CAIRO_COMPILER_TARGET_TAG https://github.com/starkware-libs/cairo.git
+ARG COMPILER_BINARY_URL
```
Comment (FabijanC) [inline Dockerfile:12]: IMO it's an unnecessary env var (i.e. build arg)

File: Dockerfile [hunk 4]
```diff
@@ -40,10 +40,8 @@ RUN apk add --no-cache libgmpxx
 
 COPY --from=cairo-lang /wheels /wheels
 
-COPY --from=compiler-builder /cairo/corelib /usr/local/bin/corelib
-COPY --from=compiler-builder /cairo/target/release/starknet-sierra-compile /usr/local/bin/target/release/starknet-sierra-compile
-COPY --from=compiler-builder /cairo/target/release/starknet-compile /usr/local/bin/target/release/starknet-compile
-COPY --from=compiler-builder /root/.local/share/scarb-install/${SCARB_VERSION}/bin/scarb /usr/local/bin/scarb
+COPY --from=compiler /cairo /usr/local/bin/cairo
```
Comment (FabijanC) [inline Dockerfile]: Won't this break compatibility with starknet-hardhat-plugin? It's expecting the binaries to be in /usr/local/bin/target/release (defined in constants.ts). Let's keep it backwards compatible, but leave a comment in this part of the file.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 5]
```diff
@@ -10,11 +9,12 @@ RUN apk add git musl-dev curl
...
+ARG ASSET_NAME
 
-RUN cargo build --release --manifest-path /cairo/crates/cairo-lang-starknet/Cargo.toml
-RUN cargo build --release --manifest-path /cairo/crates/cairo-lang-sierra-to-casm/Cargo.toml
+# Download cairo1 compiler
+RUN curl --location -O --request GET $COMPILER_BINARY_URL
+RUN tar -zxvf $ASSET_NAME
 
 # Install cairo-lang
 FROM python:3.9.13-alpine3.16 as cairo-lang
```
Comment (FabijanC) [inline Dockerfile]: Can we remove this line? Since it's the same image as on the first line

Comment (Nathan-SL) [reply Dockerfile]: https://github.com/0xSpaceShard/cairo-cli-docker/pull/19/commits/495d1553932745eef246d6013af98c9ca55441af

File: Dockerfile (hunk 4 above)
Comment (Nathan-SL) [reply Dockerfile]: https://github.com/0xSpaceShard/cairo-cli-docker/pull/19/commits/495d1553932745eef246d6013af98c9ca55441af

File: Dockerfile (hunk 3 above)
Comment (Nathan-SL) [reply Dockerfile:12]: Kept it as discussed.

File: Dockerfile (hunk 2 above)
Comment (Nathan-SL) [reply Dockerfile]: https://github.com/0xSpaceShard/cairo-cli-docker/pull/19/commits/495d1553932745eef246d6013af98c9ca55441af

File: .circleci/config.yml (hunk 1 above)
Comment (Nathan-SL) [reply .circleci/config.yml]: https://github.com/0xSpaceShard/cairo-cli-docker/pull/19/commits/f4e8f9b0916cbad645b2bc8ea5df34a4a2c7ec5b

File: Dockerfile (hunk 4 above)
Comment (FabijanC) [reply Dockerfile]: I meant a comment which explains why we are copying it to that exact directory (to keep it compatible with what the plugin is expecting).

Comment (Nathan-SL) [reply Dockerfile]: https://github.com/0xSpaceShard/cairo-cli-docker/pull/19/commits/20bb4c2bd226e897c9501273009db5888b0ea102

**Review by FabijanC — APPROVED** [review]

File: .circleci/config.yml [hunk 6]
```diff
@@ -21,24 +21,26 @@ jobs:
...
       - run:
           name: Build and push image
           command: ./build_and_push.sh
           environment:
             TAG_SUFFIX: -arm
-            CAIRO_VERSION: 0.11.2
+            CAIRO_VERSION: "v0.11.2"
             OZ_VERSION: 0.5.1
             CAIRO_COMPILER_TARGET_TAG: "v1.1.0"
+            CAIRO_COMPILER_ASSET_NAME: "release-x86_64-unknown-linux-musl.tar.gz"
```
Comment (FabijanC) [inline .circleci/config.yml:43]: Should we be using that asset for mac?

File: .circleci/config.yml [hunk 7]
```diff
@@ -21,24 +21,26 @@ jobs:
...
             SCARB_VERSION: 0.4.0
   build_push_arm:
-    machine:
-      image: ubuntu-2004:202111-02
-    resource_class: arm.medium
+    macos:
+            xcode: 14.2.0
+    resource_class: macos.m1.large.gen1
```
Comment (FabijanC) [inline .circleci/config.yml:32]: Do CircleCI's mac executors support docker?

File: .circleci/config.yml (hunk 6 above)
Comment (Nathan-SL) [reply .circleci/config.yml:43]: I think as long as we're using an alpine image for the docker build, we can't proceed with the Mac binary (`release-aarch64-apple-darwin.tar`) but I have checked by compiling cairo1 contracts and works fine.

File: .circleci/config.yml (hunk 7 above)
Comment (Nathan-SL) [reply .circleci/config.yml:32]: I think it is supported https://circleci.com/docs/using-macos/#using-multiple-executor-types-macos-docker

Comment (FabijanC) [reply .circleci/config.yml:32]: let's merge and see what happens

**Review by FabijanC — APPROVED** [review]

**Pull request merged**